File: src/code_indexing_mcp/installer/verify.py
"""Post-install checks: proof the installation works, not just that files were written.

Every step before this one reports what it *did*. These checks report what is
now *true* -- the executable runs, the launcher resolves, each harness config
still parses and names a command that exists. A wizard that says "complete" and
leaves a client silently unable to start the server is the failure mode worth
spending a few seconds to rule out.

Nothing here can fail an installation. A check that does not pass is a warning
attached to a result that already happened.
"""

from __future__ import annotations

import os
import shutil
import subprocess
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path

from . import accelerator, harnesses, shell_path
from .env_blocks import command_from_entry

# Long enough for a cold interpreter start on a loaded machine, short enough
# that a hung executable does not hold the wizard open.
HELP_TIMEOUT_SECONDS = 30.0


@dataclass(frozen=True)
class Check:
    name: str
    status: str  # "ok" | "warn" | "fail"
    detail: str = ""

    @property
    def ok(self) -> bool:
        return self.status == "ok"


def _server_runs(install_directory: Path) -> Check:
    executable = accelerator.server_executable(install_directory)
    if not executable.is_file():
        return Check("server executable", "fail", f"missing: {executable}")
    try:
        completed = subprocess.run(
            [str(executable), "--help"],
            capture_output=True,
            timeout=HELP_TIMEOUT_SECONDS,
            check=False,
        )
    except OSError as exc:
        return Check("server executable", "fail", f"{executable} could not be launched: {exc}")
    except subprocess.TimeoutExpired:
        return Check(
            "server executable",
            "fail",
            f"{executable} did not answer --help within {HELP_TIMEOUT_SECONDS:.0f}s",
        )
    if completed.returncode != 0:
        message = completed.stderr.decode("utf-8", "replace").strip().splitlines()
        return Check(
            "server executable",
            "fail",
            f"{executable} --help exited {completed.returncode}: {message[-1] if message else ''}",
        )
    return Check("server executable", "ok", str(executable))


def _launcher_resolves(
    launcher: shell_path.LauncherResult | None,
    profiles_updated: Sequence[Path],
    *,
    environment: Mapping[str, str] | None = None,
) -> Check:
    environment = os.environ if environment is None else environment
    if launcher is None:
        return Check("command on PATH", "warn", "no launcher was requested")
    if not launcher.ok:
        return Check("command on PATH", "warn", f"launcher not created: {launcher.detail}")
    found = shutil.which(shell_path.LAUNCHER_NAME, path=environment.get("PATH", ""))
    if found is not None and Path(found).resolve() == launcher.path.resolve():
        return Check("command on PATH", "ok", found)
    if profiles_updated:
        # The block was written into a profile this shell read before it existed.
        # That is the expected state right after an install, not a problem. The
        # hint has to come from the environment being verified, not this process:
        # they differ whenever the caller passed one in.
        hint = shell_path.activation_hint(profiles_updated, environment=environment)
        return Check("command on PATH", "warn", f"resolves once you start a new shell ({hint})")
    if found is not None:
        return Check("command on PATH", "warn", f"{found} is found first, not {launcher.path}")
    return Check(
        "command on PATH",
        "warn",
        f"{launcher.path.parent} is not on PATH; add it to use the command by name",
    )


def _harness_entries(
    configured: Sequence[tuple[str, Path]],
    install_directory: Path,
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> list[Check]:
    """Re-read what was just written: a merge that parsed can still land wrong."""

    expected = accelerator.server_executable(install_directory)
    checks: list[Check] = []
    for slug, path in configured:
        name = f"{slug} configuration"
        try:
            entry = harnesses.read_server_entry(slug, home=home, environment=environment)
        except Exception as exc:  # a harness path this platform cannot even name
            checks.append(Check(name, "warn", f"could not be re-read: {exc}"))
            continue
        if entry is None:
            checks.append(Check(name, "warn", f"no server entry found in {path} after writing it"))
            continue
        command = command_from_entry(slug, entry)
        if command is None:
            checks.append(Check(name, "warn", f"the entry in {path} names no command"))
        elif Path(command) != expected:
            checks.append(Check(name, "warn", f"the entry names {command}, expected {expected}"))
        else:
            checks.append(Check(name, "ok", str(path)))
    return checks


def _accelerator_recorded(install_directory: Path) -> Check:
    prepared = accelerator.prepared_accelerator(install_directory)
    if prepared is None:
        return Check("accelerator record", "warn", "no record; the server will resolve at startup")
    return Check("accelerator record", "ok", prepared)


def _skill_links(
    slugs: Sequence[str],
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> list[Check]:
    checks: list[Check] = []
    for slug in slugs:
        directory = harnesses.skill_directory(slug, home=home, environment=environment)
        if directory is None or not directory.is_dir():
            continue
        # Only the links this installer made. A broken symlink the user put in
        # their own skill directory is not this installation's to report on.
        broken = [
            entry.name
            for entry in directory.iterdir()
            if harnesses.is_bundled_skill_link(entry) and not entry.resolve().is_dir()
        ]
        name = f"{slug} skills"
        if broken:
            checks.append(Check(name, "warn", f"links point nowhere: {', '.join(sorted(broken))}"))
        else:
            checks.append(Check(name, "ok", str(directory)))
    return checks


def run_checks(
    install_directory: Path,
    configured: Sequence[tuple[str, Path]] = (),
    *,
    launcher: shell_path.LauncherResult | None = None,
    profiles_updated: Sequence[Path] = (),
    accelerator_was_prepared: bool = True,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> tuple[Check, ...]:
    """Check the finished installation. Never raises; problems come back as warnings."""

    checks = [_server_runs(install_directory)]
    checks.append(_launcher_resolves(launcher, profiles_updated, environment=environment))
    checks.extend(
        _harness_entries(configured, install_directory, home=home, environment=environment)
    )
    if accelerator_was_prepared:
        checks.append(_accelerator_recorded(install_directory))
    checks.extend(
        _skill_links([slug for slug, _ in configured], home=home, environment=environment)
    )
    return tuple(checks)


def format_check(check: Check) -> str:
    """One display line per check.

    No square brackets: this string is handed to Rich-rendered widgets, which
    would read ``[ok]`` as a markup tag and swallow it.
    """

    marker = {"ok": "ok  ", "warn": "warn", "fail": "FAIL"}[check.status]
    line = f"{marker} - {check.name}"
    return f"{line}: {check.detail}" if check.detail else line

File: ts/packages/installer/src/verify.ts
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { preparedAccelerator, serverExecutable } from "./accelerator.ts";
import { commandFromEntry } from "./env-blocks.ts";
import { isBundledSkillLink, readServerEntry, skillDirectory } from "./harnesses.ts";
import {
  activationHint,
  LAUNCHER_NAME,
  type LauncherResult,
  launcherOk,
  launcherPath,
} from "./shell-path.ts";

export const HELP_TIMEOUT_SECONDS = 30;

export interface Check {
  readonly name: string;
  readonly status: "ok" | "warn" | "fail";
  readonly detail: string;
  readonly ok: boolean;
}

function check(name: string, status: Check["status"], detail = ""): Check {
  return { name, status, detail, ok: status === "ok" };
}

function whichOnPath(name: string, pathValue: string): string | null {
  for (const directory of pathValue.split(path.delimiter)) {
    if (directory === "") continue;
    for (const candidate of [name, `${name}.cmd`, `${name}.exe`]) {
      const filePath = path.join(directory, candidate);
      try {
        if (fs.statSync(filePath).isFile()) return filePath;
      } catch {}
    }
  }
  return null;
}

function serverRuns(installDirectory: string): Check {
  const executable = serverExecutable(installDirectory);
  try {
    if (!fs.statSync(executable).isFile()) {
      return check("server executable", "fail", `missing: ${executable}`);
    }
  } catch {
    return check("server executable", "fail", `missing: ${executable}`);
  }
  const completed = spawnSync(executable, ["--help"], {
    encoding: "utf8",
    timeout: HELP_TIMEOUT_SECONDS * 1000,
  });
  if (completed.error !== undefined) {
    if (completed.error.message.includes("ETIMEDOUT") || completed.signal === "SIGTERM") {
      return check(
        "server executable",
        "fail",
        `${executable} did not answer --help within ${HELP_TIMEOUT_SECONDS.toFixed(0)}s`,
      );
    }
    return check(
      "server executable",
      "fail",
      `${executable} could not be launched: ${completed.error.message}`,
    );
  }
  if (completed.status !== 0) {
    const message = (completed.stderr ?? "").trim().split(/\r?\n/);
    return check(
      "server executable",
      "fail",
      `${executable} --help exited ${completed.status}: ${message[message.length - 1] ?? ""}`,
    );
  }
  return check("server executable", "ok", executable);
}

function launcherResolves(
  launcher: LauncherResult | null,
  profilesUpdated: readonly string[],
  environment: NodeJS.ProcessEnv,
): Check {
  if (launcher === null) return check("command on PATH", "warn", "no launcher was requested");
  if (!launcherOk(launcher)) {
    return check("command on PATH", "warn", `launcher not created: ${launcher.detail}`);
  }
  const found = whichOnPath(LAUNCHER_NAME, environment.PATH ?? environment.Path ?? "");
  if (found !== null) {
    try {
      if (fs.realpathSync(found) === fs.realpathSync(launcher.path)) {
        return check("command on PATH", "ok", found);
      }
    } catch {
      // Fall through.
    }
  }
  if (profilesUpdated.length > 0) {
    const hint = activationHint(profilesUpdated, { environment });
    return check("command on PATH", "warn", `resolves once you start a new shell (${hint})`);
  }
  if (found !== null) {
    return check("command on PATH", "warn", `${found} is found first, not ${launcher.path}`);
  }
  return check(
    "command on PATH",
    "warn",
    `${path.dirname(launcher.path)} is not on PATH; add it to use the command by name`,
  );
}

function harnessEntries(
  configured: readonly [string, string][],
  installDirectory: string,
  options: { home?: string; environment?: NodeJS.ProcessEnv },
): Check[] {
  const expected = serverExecutable(installDirectory);
  const checks: Check[] = [];
  for (const [slug, filePath] of configured) {
    const name = `${slug} configuration`;
    let entry: Record<string, unknown> | null;
    try {
      entry = readServerEntry(slug, options);
    } catch (error) {
      checks.push(
        check(
          name,
          "warn",
          `could not be re-read: ${error instanceof Error ? error.message : String(error)}`,
        ),
      );
      continue;
    }
    if (entry === null) {
      checks.push(check(name, "warn", `no server entry found in ${filePath} after writing it`));
      continue;
    }
    const command = commandFromEntry(slug, entry);
    if (command === null)
      checks.push(check(name, "warn", `the entry in ${filePath} names no command`));
    else if (command !== expected) {
      checks.push(check(name, "warn", `the entry names ${command}, expected ${expected}`));
    } else {
      checks.push(check(name, "ok", filePath));
    }
  }
  return checks;
}

function acceleratorRecorded(): Check {
  const prepared = preparedAccelerator("");
  if (prepared === null) {
    return check("accelerator record", "warn", "no record; the server will resolve at startup");
  }
  return check("accelerator record", "ok", prepared);
}

function skillLinks(
  slugs: readonly string[],
  options: { home?: string; environment?: NodeJS.ProcessEnv },
): Check[] {
  const checks: Check[] = [];
  for (const slug of slugs) {
    const directory = skillDirectory(slug, options);
    if (directory === null || !fs.existsSync(directory) || !fs.statSync(directory).isDirectory()) {
      continue;
    }
    const broken = fs
      .readdirSync(directory)
      .map((name) => path.join(directory, name))
      .filter((entry) => {
        if (!isBundledSkillLink(entry)) return false;
        try {
          return !fs.statSync(fs.realpathSync(entry)).isDirectory();
        } catch {
          return true;
        }
      })
      .map((entry) => path.basename(entry));
    const name = `${slug} skills`;
    if (broken.length > 0) {
      checks.push(check(name, "warn", `links point nowhere: ${broken.sort().join(", ")}`));
    } else {
      checks.push(check(name, "ok", directory));
    }
  }
  return checks;
}

export function runChecks(
  installDirectory: string,
  configured: readonly [string, string][] = [],
  options: {
    launcher?: LauncherResult | null;
    profilesUpdated?: readonly string[];
    acceleratorWasPrepared?: boolean;
    home?: string;
    environment?: NodeJS.ProcessEnv;
  } = {},
): Check[] {
  const environment = options.environment ?? process.env;
  const checks = [serverRuns(installDirectory)];
  checks.push(
    launcherResolves(options.launcher ?? null, options.profilesUpdated ?? [], environment),
  );
  checks.push(...harnessEntries(configured, installDirectory, options));
  if (options.acceleratorWasPrepared !== false) checks.push(acceleratorRecorded());
  checks.push(
    ...skillLinks(
      configured.map(([slug]) => slug),
      options,
    ),
  );
  return checks;
}

export function runUpdateChecks(
  installDirectory: string,
  configured: readonly [string, string][] = [],
  options: {
    acceleratorWasPrepared?: boolean;
    home?: string;
    environment?: NodeJS.ProcessEnv;
  } = {},
): Check[] {
  const checks = [serverRuns(installDirectory)];
  checks.push(...harnessEntries(configured, installDirectory, options));
  if (options.acceleratorWasPrepared !== false) checks.push(acceleratorRecorded());
  checks.push(
    ...skillLinks(
      configured.map(([slug]) => slug),
      options,
    ),
  );
  return checks;
}

export function formatCheck(item: Check): string {
  const marker = { ok: "ok  ", warn: "warn", fail: "FAIL" }[item.status];
  const line = `${marker} - ${item.name}`;
  return item.detail !== "" ? `${line}: ${item.detail}` : line;
}

export { launcherPath };
